hail/python/hail/ir/renderer.py

```diff
@@ -134,3 +135,563 @@ def __call__(self, x: 'Renderable'):
+                out_builder.extend(let_body)
+            out_builder.extend(self.builder)
+            num_lets = len(let_bodies)
+            for _ in range(num_lets):
+                out_builder.append(')')
+
+        def make_child_frame(self,
+                             renderer: 'CSERenderer',
+                             binding_sites: Dict[int, BindingSite],
+                             builder: Sequence[str],
+                             bindings_stack: 'Sequence[CSEPrintPass.BindingsStackFrame]',
+                             min_binding_depth: int,
+                             min_value_binding_depth: int,
+                             scan_scope: bool,
+                             depth: int):
+            child = self.children[self.child_idx]
+            return self.make(child, renderer, binding_sites, builder, bindings_stack,
+                             min_binding_depth, min_value_binding_depth,
+                             scan_scope, depth)
+
+        @staticmethod
+        def make(node: Renderable,
+                 renderer: 'CSERenderer',
+                 binding_sites: Dict[int, BindingSite],
+                 builder: Sequence[str],
+                 bindings_stack: 'Sequence[CSEPrintPass.BindingsStackFrame]',
+                 min_binding_depth: int,
+                 min_value_binding_depth: int,
+                 scan_scope: bool,
+                 depth: int):
+            insert_lets = (id(node) in binding_sites
+                           and depth == binding_sites[id(node)].depth
+                           and (len(binding_sites[id(node)].lifted_lets) > 0
+                                or len(binding_sites[id(node)].agg_lifted_lets) > 0))
+            state = CSEPrintPass.StackFrame(node,
+                                            node.render_children(renderer),
+                                            builder,
+                                            min_binding_depth,
+                                            min_value_binding_depth,
+                                            scan_scope, depth, insert_lets)
+            if insert_lets:
+                state.builder = []
+                bindings_stack.append(
+                    CSEPrintPass.StackFrame.make_bindings_stack_frame(
+                        binding_sites[id(node)]))
+            head = node.render_head(renderer)
+            if head != '':
+                state.builder.append(head)
+            return state
+
+        @staticmethod
+        def make_bindings_stack_frame(site: BindingSite):
+            return CSEPrintPass.BindingsStackFrame(
+                depth=site.depth,
+                lifted_lets=site.lifted_lets,
+                agg_lifted_lets=site.agg_lifted_lets,
+                scan_lifted_lets=site.scan_lifted_lets,
+                visited={},
+                agg_visited={},
+                scan_visited={},
+                let_bodies=[])
+
+
+class CSERenderer:
```

Comment:
Renderer should be an abstract base class, with [Stupid]Renderer and CSERenderer extending it. Currently the design works because you define the right methods, but the types don't work out.